File: src/WirePacker.cpp
/**
 * Modified based on https://github.com/gutierrezps/ESP32_I2C_Slave
 */

#include "WirePacker.h"

#include "WireCrc.h"

WirePacker::WirePacker(uint32_t bufferSize) : bufferSize_(bufferSize) {
    buffer_ = (uint8_t *)malloc(sizeof(uint8_t) * bufferSize_);
    if (buffer_ == nullptr) {
        bufferSize_ = 0;
        return;
    }
    reset();
}

WirePacker::~WirePacker() {
    free(buffer_);
    buffer_ = nullptr;
}

size_t WirePacker::write(uint8_t data) {
    if (!isPacketOpen_) {
        return 0;
    }

    // leave room for crc and end bytes
    if (totalLength_ >= bufferSize_ - 2) {
        return 0;
    }

    buffer_[index_] = data;
    ++index_;
    totalLength_ = index_;
    return 1;
}

size_t WirePacker::write(const uint8_t *data, size_t quantity) {
    for (size_t i = 0; i < quantity; ++i) {
        if (!write(data[i])) {
            return i;
        }
    }
    return quantity;
}

void WirePacker::end() {
    isPacketOpen_ = false;

    // make room for CRC byte
    ++index_;

    buffer_[index_] = frameEnd_;
    ++index_;
    totalLength_ = index_;

    int offset;
    if (bufferSize_ < (1 << 8)) {
        buffer_[2] = totalLength_ & 0xFF;
        offset = 3;
    } else if (bufferSize_ < (1 << 16)) {
        buffer_[2] = (totalLength_ >> 8) & 0xFF;
        buffer_[3] = totalLength_ & 0xFF;
        offset = 4;
    } else {
        buffer_[2] = (totalLength_ >> 16) & 0xFF;
        buffer_[3] = (totalLength_ >> 8) & 0xFF;
        buffer_[4] = totalLength_ & 0xFF;
        offset = 5;
    }

    uint32_t payloadLength = totalLength_ - ignoreLength_;

    WireCrc crc8;
    uint8_t crc = crc8.update(buffer_ + offset, payloadLength);
    buffer_[index_ - 2] = crc;

    // prepare for reading
    index_ = 0;
}

size_t WirePacker::available() {
    if (isPacketOpen_) {
        return 0;
    }

    return totalLength_ - index_;
}

int WirePacker::read() {
    int value = -1;

    if (!isPacketOpen_ && index_ < totalLength_) {
        value = buffer_[index_];
        ++index_;
    }

    return value;
}

void WirePacker::reset() {
    buffer_[0] = frameStart_;
    // ignore start, length, crc and end bytes
    if (bufferSize_ < (1 << 8)) {
        buffer_[1] = 1;
        index_ = 3;
        ignoreLength_ = 5;
    } else if (bufferSize_ < (1 << 16)) {
        buffer_[1] = 2;
        index_ = 4;
        ignoreLength_ = 6;
    } else {
        buffer_[1] = 3;
        index_ = 5;
        ignoreLength_ = 7;
    }
    totalLength_ = index_;
    isPacketOpen_ = true;
}

File: i2c_for_esp32/wirepacker.py
from i2c_for_esp32.wirecrc import WireCrc


class WirePacker(object):

    def __init__(self, buffer_size=5):
        self.buffer_size = buffer_size
        self.frameStart = 0x02
        self.frameEnd = 0x04
        self.buffer = bytearray(self.buffer_size)
        self.index = 0
        self.totalLength = 0
        self.isPacketOpen = False
        self.reset()

    def write(self, data):
        if not self.isPacketOpen:
            return 0
        if self.totalLength >= self.buffer_size - 2:
            return 0
        self.buffer[self.index] = data
        self.index += 1
        self.totalLength = self.index
        return 1

    def end(self):
        self.isPacketOpen = False
        self.index += 1
        self.buffer[self.index] = self.frameEnd
        self.index += 1
        self.totalLength = self.index

        if self.buffer_size < (1 << 8):
            self.buffer[2] = self.totalLength & 0xFF
            offset = 3
        elif self.buffer_size < (1 << 16):
            self.buffer[2] = (self.totalLength >> 8) & 0xFF
            self.buffer[3] = self.totalLength & 0xFF
            offset = 4
        else:
            self.buffer[2] = (self.totalLength >> 16) & 0xFF
            self.buffer[3] = (self.totalLength >> 8) & 0xFF
            self.buffer[4] = self.totalLength & 0xFF
            offset = 5

        payloadLength = self.totalLength - self.ignoreLength

        crc8 = WireCrc()
        crc = crc8.update(self.buffer[offset:], payloadLength)
        self.buffer[self.index - 2] = crc
        self.index = 0

    def available(self):
        if self.isPacketOpen:
            return 0
        return self.totalLength - self.index

    def read(self):
        if self.isPacketOpen or self.index >= self.totalLength:
            return -1
        value = self.buffer[self.index]
        self.index += 1
        return value

    def reset(self):
        self.buffer[0] = self.frameStart
        if self.buffer_size < (1 << 8):
            self.buffer[1] = 1
            self.index = 3
            self.ignoreLength = 5
        elif self.buffer_size < (1 << 16):
            self.buffer[1] = 2
            self.index = 4
            self.ignoreLength = 6
        else:
            self.buffer[1] = 3
            self.index = 5
            self.ignoreLength = 7
        self.totalLength = self.index
        self.isPacketOpen = True
